### DISCUSSION TITLE
Who uses objection.js in production?

### DISCUSSION THREAD

**Discussion by jzimmek** [Vincit/objection.js#2464] [Show and tell]
Hi there, I am currently evaluating a bunch of orm-a-like libraries for an upcoming project.

I almost gave up after digging deeper into sequelize and typeorm .

Having a PostgreSQL, java (hibernate) and ruby (ar/sequel) background, I was kinda disappointed by nodejs orm libraries. They are not bad, but feel very immature in comparison to java/ruby world.

Then, by accident, I found objectionjs while reading the knexjs README.

Objectionjs is exactly my way of thinking. Pragmatic but not to abstract. Sequelize/Typeorm might have more github stars/forks, but comparing them API- as well as Issue- and feature-wise, objectionjs feels like a way better fit to me (and probably others who also have a relational sql db background).

It deserves more traction and I am asking myself - why is objectionjs not more popular? And who is actually using it in production?

I know it sounds stupid, but is there some who-uses-it listing I can show to an IT decision maker?

Nevertheless, you did an awesome job and built a great library!

**Comment by kibertoad** [comment]
Companies that I am aware of:
https://www.shift4.com/
https://www.bloomon.nl/

**Comment by devinivy** [comment]
We adopted it as our de facto SQL ORM at [Big Room Studios](https://www.bigroomstudios.com/) (you can probably even spot an objection model on the homepage!), and our development team has only been very happy with it, having used a variety of ORMs.  The size of the community is actually very comfortable—it's large enough that you have active participation on issues, good community support in gitter, and regular contributions, but small enough that maintainers have been able to effectively handle issues without generating a massive backlog.  It's also built upon knex (and there's overlap in contributors between objection and knex), which is _very_ widely used.  Lastly, it's also the officially supported ORM of @hapipal.

**Comment by kibertoad** [comment]
Note that Objection.js is NOT an ORM in a sense you are used to in Hibernate world - it is far more lightweight, it would be more correct to think of it as a query builder that also takes care of relation/upsertion/validation logic for you. It doesn't have out-of-the-box locking. It doesn't have transparent lazy loading. It doesn't have DB schema generation from models. If you need all of these - TypeORM or Sequelize would be a better fit for you, but if you don't need them or are OK with doing them yourself - Objection.js is by far a better option. It is predictable and never gets into your way.

**Comment by koskimas** [comment]
Objection has been used in various (10 -20) projects at Vincit even before it was released as open source. I can't give any names as I'm not sure which customers are public references for Vincit (I don't work there anymore). Maybe @elhigu can provide some kind of list?

**Comment by elhigu** [comment]
> Maybe @elhigu can provide some kind of list?
 
I’m trying to get at least current number of companies + sizes in case we cannot give the names.

**Comment by mitchellporter** [comment]
@jzimmek you might also want to ask this on the node subreddit: [reddit.com/r/node](https://reddit.com/r/node)

There have been many mentions of objection on there that's how I first found out about it.

**Comment by flipace** [comment]
[We're](https://ovos.at/en/) (company of ~30+) using it pretty heavily since about a year now.
Previously used bookshelf, and after being burned by it at some point we looked into alternatives.
TypeORM looked really promising, but after some digging we hit some dead ends.

objection.js is really pleasent to use and I believe that everyone on our team (~10+) feels very comfortable with it. Documentation is top-notch and the community, even though it might not be as big as others, is extremely helpful. Also, the codebase itself is pretty clear and relatively easy to understand in comparison to other libs.

(thank you - to the creators and maintainers of objection.js - you make our work a lot more fun)

**Comment by elhigu** [comment]
Vincit (software contractor company of ~500 empoyees) is usually using objection.js for Node projects, which needs an ORM. 

I'm still working on rough number of our clients who are using it (I haven't been at the office lately and I don't have access to check out client data myself).

**Comment by hibearpanda** [comment]
We're a small startup in Asia - been using objection since we first developed our product.

**Comment by heisian** [comment]
Paul Jensen - London Node User Group
https://www.youtube.com/watch?v=RyQ1MTVjYK8

**Comment by elhigu** [comment]
That would be nice link to add to readme or something :)

**Comment by kylealwyn** [comment]
We do! We're a health tech company doing online doctor visits in the US: https://www.heydoctor.co/

**Comment by capaj** [comment]
![looop-logo](https://www.looop.co/wp-content/themes/looop2017B/img/looop-logo-green.svg)
We're a small remote startup(5 devs ATM) called [Looop](https://www.looop.co/).
We rewrote our monolith app from Sequelize to objection.js and so far it has been a godsend. We have 76 tables in postgre so relation expressions are probably our most used feature. We still have like 15% of our API on sequelize, but every day we rip out more and more. BTW we expose our objection.js models to the consumers just by decorating them with [decapi](https://github.com/capaj/decapi). Which is quite a unique way of doing that I'd say.

**Comment by bryanbrunt** [comment]
I recently rewrote a version of the board game Risk using objection. I can't even begin to describe how much simpler it was to use objection.js than the older java bean code. I probably have 70% less code.

And when comparing objection to sequelize -- why on earth would you subject yourself to the added complexity of that tool?

Seriously, sequelize is bloated. And don't even get me started about sails and typeorm. I was about to give up on using node.js as the backend for the game, and then I found objection.js.

Bookshelf is a step in the right direction, but objection seems like a refactored, simplified bookshelf.

The game is going to be released as a Steam Early Release here in the next week:

http://www.denizenseven.com/

https://store.steampowered.com/app/952090/Grand_Strategy/

https://play.google.com/store/apps/details?id=org.grandstrategy.client

**Comment by joaonice** [comment]
We're a Portuguese tech company (company of ~30) that proudly uses `objection.js` in several projects that are currently running on production. 

https://seegno.com

![4e22b6b0c868e66b04d18f4985380254](https://user-images.githubusercontent.com/1272480/47013193-3d371e00-d13e-11e8-8d3e-a087ee3c64d7.jpg)

**Comment by heisian** [comment]
![screen shot 2018-10-16 at 10 36 54 am](https://user-images.githubusercontent.com/4141828/47035657-6ef4b880-d12f-11e8-8cfa-14b62867b270.png)

Ethos Technologies, Inc. - We offer term life insurance for citizens and residents in the U.S.

When we migrated from MongoDB to Postgres, we tried several ORMs and Objection.js stood out to us the most - it is one of the most solid, well-architected packages in our stack - can't recommend it enough!

[Ethos - Insure Your Life](https://getethos.com)

**Comment by kapouer** [comment]
I am ! Though you can only see the front-end right now, the back-end being not polished enough.
99% of the project being open source, a typical example being how json schemas are used to build an ajv validator for objection:
[here some schema](https://github.com/pageboard/elements/blob/9610221b1b936d695e5dcf6dce13e77f3cfc386d/elements/page.js#L8)
[here how objection/ajv is setup](https://github.com/pageboard/api/blob/3aad1caf79679314619922344eeeaadf38a3a477/models/block.js#L137)
[here some front-end](https://laetitia-casta.fr)
It's a modern (well, though i'm not really the best programmer of the world, i can say the concepts are way on top of the cake) website editor i've been building upon objection and prosemirror and DOM.
Soon to be available for all !

**Comment by dearsaturn** [comment]
We do at Social Native.

**Comment by johannessjoberg** [comment]
![dporganizer-primary-logo-rgb](https://user-images.githubusercontent.com/4087299/48255613-8ad93a80-e40d-11e8-8875-c84c40d4b3eb.png)

We, [DPOrganizer](https://www.dporganizer.com/), are a Swedish startup with about ~20 developers. We choose `objection.js` for a new service after evaluating a bunch of different Node.js ORMs about a year ago. We have not looked back since and will continue to invest in this great lib, both as consumers and contributors.

`objection.js` itself, the maintainers and the surrounding community has been great to work with. A big thanks to everyone involved and keep up the awesome work!

**Comment by o8e** [comment]
We've been using it at [Cazana](https://cazana.com) for 2+ years.

**Comment by rictorres** [comment]
when i was at https://mycujoo.tv we started experimenting with objection in late 2016. it went to production in early 2017 (i no longer work there). it's used in one of the largest backend services (thousands of reqs/s)

**Comment by kelgendy1204** [comment]
As a personal experience in a side project when I searched what ORM should i use, I Find the most popular one which is sequelize, after about a month of working I suffered a lot, its documentation isn't clear, strange way of coding (at least for me).
I decided to switch to objection after a lot of searching and comparing.
Oh my god, I made a migration of weeks of work with sequelize in just a few days.
Objection is such a great ORM, easy and clear documentation, predictable way of coding. In my opinion it is like eloquent for php.

**Comment by AnianZ** [comment]
At [cioplenu](https://cioplenu.com) we also switched from bookshelf to objection.js in the very early stages of our product. We are still a fairly small company but we have been very happy with that choice. The clear and concise design with excellent support for ES6 or Typescript, the option to easily fall back to knex when needed, the great compatibility with PosgreSQL, the great community and documentation really sets it apart.

**Comment by jordaaash** [comment]
At [Stylisted](https://stylisted.com), Objection powers our entire API.

- The way `QueryBuilder` is [exposed for subclassing](http://vincit.github.io/objection.js/#custom-query-builder) by models (or using HOCs) makes for many reusable patterns between models while allowing for idiosyncrasies among them
- [`onBuild`](http://vincit.github.io/objection.js/#onbuild)/[`onBuildKnex`](http://vincit.github.io/objection.js/#onbuildknex)/`onAfter2` (private API) operations provide low-level mechanisms to dynamically manipulate queries and results
- Lifecycle methods make async side effects (transactional or not) very simple
- Query [context](http://vincit.github.io/objection.js/#context) is great for authorization, filtering, and API request-specific behaviors that models should be aware of
- [Relation mappings](http://vincit.github.io/objection.js/#relationmappings) are very flexible -- it's straightforward to create relations from a model to itself for complex self-joins, for example
- [Composite keys](http://vincit.github.io/objection.js/#composite-keys) are handled correctly throughout the library
- Model JSON schemas are a nice way of describing and validating inputs/outputs at many levels of abstraction (we use them in API resources now too), but you do need to get familiar with AJV's many options
- [Relation expressions](http://vincit.github.io/objection.js/#relationexpression) with named filters, `allowEager`/`eager`/`joinEager`, and `allowInsert`/`insertGraph` make efficiently fetching/creating complex objects via single requests a snap

There's a lot more, but the point is this: the design of the library is elegant and intuitive. I've found it very easy to write code that I think will work based on the primitives provided and it just does. The source code is also easy to read, modify, and test. If you've come from Bookshelf, as we did, you'll know that none of those things hold true.

Objection gets lots of little things right by separating query building, data modeling, and validation, and doesn't encourage Rails-y patterns like polymorphism. This really reduces the O/R impedance mismatch that you experience from Bookshelf and many other ORMs.

**Comment by theogravity** [comment]
![image](https://user-images.githubusercontent.com/855434/52824461-84daa180-306d-11e9-8022-1b51b7bdad0b.png)

We use it for most of our node.js projects at Viv Labs (we're a Samsung company). We write the [Bixby AI assistant](https://www.samsung.com/us/explore/bixby/) found in many Samsung devices, like the Galaxy phones.

Objection.js is heavily used for https://bixbydevelopers.com, a place where you can learn how to build extensions to Bixby called "Capsules", (much like Alexa skills).

Since we use GraphQL, the most powerful feature that objection offers for this is the `graphInsert` / `graphUpsert` methods, which play extremely well into how GraphQL passes structured data and how we can use objection's graph methods to update our database.

I really like that it uses `ajv` as its validation backend, which makes sharing the models / schemas themselves cross-compatible with any of our other services that might want to do validation using our models (since `ajv` is `json-schema` spec compliant and objection also uses the spec to define the model), but do not require db access.

Another plus is how it's decoupled from the DBAL (database abstraction layer), where you use `knex` (or a knex-compatible API) to fill in that piece, which makes it very lightweight.

Overall, I recommend it to everyone who is looking for a lightweight ORM.

**Comment by tnolet** [comment]
I use Objection since version 1.4 on https://checklyhq.com, a monitoring SaaS. I use it for the reasons mentioned above. Feel free to use the logo or screenshot.

![image](https://user-images.githubusercontent.com/3802923/56095778-86470f80-5ee0-11e9-90f1-40eab58f4da2.png)

![logo_script_racoon_horizontal_795](https://user-images.githubusercontent.com/3802923/56095798-a971bf00-5ee0-11e9-996d-72a1786c817f.png)

**Comment by brightrain** [comment]
![image](https://user-images.githubusercontent.com/1395985/56940587-a247d500-6acc-11e9-97cc-37962c83cefe.png)

We use it in a crowd sourced mapping application called Map Our Borders, it's perfect for our needs.

[mapourborders.com](url)

BIG Thanks!

**Comment by moige01** [comment]
Hi,

The Dev Team of Surnet Telecom, which is an company that maintains and develop the payment gateway for the 90% of the Venezuelan Airlines, are actually using Objection.js in production environment with goods results.

**Comment by ais-one** [comment]
Hi,

I am currently working on a project with a Fund in Asia.

We are using ObjectionJS in production.

I use it for NodeJS projects that require a connection to the supported SQLs.

It may be a little too much to ask If it could do Mongo DB as well :)

**Comment by broom9** [comment]
We ([Flux Work](http://fluxwork.co)) are using Objection.js in production in a Node/Express-powered API server. It has been pleasant to use. Thanks for sharing such a great library!

**Comment by cludwig28** [comment]
https://www.navistone.com/ , a company of ~25. We use Objection.js in production for our lightweight node.js APIs. Superb documentation and easy to use.

**Comment by jitterfunc** [comment]
I think this should not remain as an issue, please create a page in the wiki for "production users" citing these references, so that it's easily findable and later on other users can link their organizations as a pull request if they want to.

**Comment by jitterfunc** [comment]
I have created a wiki page for objection.js consolidating the information provided in this issue till now.
[Companies using objection.js in production](https://github.com/Vincit/objection.js/wiki/Companies-using-objection.js-in-production).

**Comment by youmustfight** [comment]
[Upsolve](https://upsolve.org) has been using it with the graphql plugin and we love it! Custom query builders are a huge part of our API's design now
![image](https://user-images.githubusercontent.com/4956240/71206764-57edb100-2273-11ea-89b9-818fa3fd678b.png)

**Comment by feross** [comment]
I used `objection` to build [BitMidi](https://bitmidi.com), a free MIDI music database that gets millions of pageviews per month. I plan to use it on future projects. It's been a joy to work with.

[![bitmidi](https://user-images.githubusercontent.com/121766/71537754-96b7e280-28d5-11ea-81e6-710ea37ddcf0.png)](https://bitmidi.com)

**Comment by globsecure** [comment]
@johannessjoberg -> Sorry to bother you at this time, but I would like to check if your view is still the same today.

**Comment by johannessjoberg** [comment]
@globsecure No problem, and yes, I still have the same view. I think that objection simplifies and minimizes the code concerning relationships without getting in the way and provides a very powerful query builder that I enjoy working with. This blog post https://www.jakso.me/blog/objection-to-orm-hatred by @koskimas resonates well with my view on ORMs and working with SQL.

Edit: Another thing that I highly value is the flexibility provided by that possibility to easily switch to raw SQL using `knex.raw` if needed.

**Comment by weylermaldonado** [comment]
I used `objection` to build the [Voxia's](https://voxia.soldai.com/panel/) API, and all the products of my work has been built used `objection`.

**Comment by crespowang** [comment]
We've been with ObjectionJs since inquisitive.com was launched two years ago, very happy with it.

**Comment by ianldgs** [comment]
We at [deedmob.com](https://deedmob.com) are in the process of migrating from Bookshelf.

**Comment by kennethacohen** [comment]
Just moved from Sequelize to Objection for [launchbot.app](http://launchbot.app/). So. Much. Better.

**Comment by vigzmv** [comment]
We at [Novelship](https://novelship.com) have been using and loving objection for more than 2.5 years now.
~We want to upgrade to v2 and but are stuck with v1 because we rely on [objection-find](https://github.com/Vincit/objection-find) which doesn't support v2 yet~

Update:

Objection-find got an upgrade, supports objection@2 now!

**Comment by koskimas** [comment]
I deleted @ofeenee's comment because it wasn't relevant to this thread. Feel free to open a separate issue. Let's keep this issue purely as a list of companies/projects that use objection.

**Comment by roggervalf** [comment]
I'm currently working on one of Autodesk's project and they are using this amazing package. Since I used this tool is pretty easy to understand and use.

**Comment by ndevvy** [comment]
we're using Objection at [substack](https://substack.com)!

**Comment by davisml** [comment]
We're using Objection at [Pinkaloo](https://www.pinkaloo.com) 💓

**Comment by acro5piano** [comment]
We're using Objection at [Remotehour](https://remotehour.com?utm_campaign=objection.js) with GraphQL.

**Comment by alex-all3dp** [comment]
We're using objection.js at [Craftcloud](https://craftcloud3d.com/)

<img width="380" alt="logo-screenshot" src="https://user-images.githubusercontent.com/52403795/111127652-5a24e500-8574-11eb-903d-6b52dc795f84.png">

**Comment by ehubbell** [comment]
We're using Objection at [Blush](https://blush-now.com) and a couple other private projects.

**Comment by alexisohayon** [comment]
Proudly using it too at [CitizenPlane](https://citizenplane.com)! We're ~20+ company, 7 techs, Objection in production on all services that need to query our db.
Best SQL ORM for Node we've had so far :)

**Comment by laznic** [comment]
Using Objection at [Pizza Cartel](https://pizzacartel.fi) and [eggwp](https://eggwp.co)!

**Comment by Haisum92** [comment]
I am connected with a company who is using it for [VIOP](https://searchunifiedcommunications.techtarget.com/definition/VoIP).

**Comment by feross** [comment]
We use objection.js on [Wormhole](https://wormhole.app).

**Comment by laurisvan** [comment]
Our six person startup, [Rescomms](https://www.rescomms.com/), has been using Objection.js for its whole three year existence. Currently we are serving some 0,5M heavy GraphQL queries/day.

I must say having a proper query builder that can adapt to the varying GraphQL payloads and inputs has solved a lot of problems to us. Until you face the problem of adapting to arbitrary inputs and outputs, you might not know you need a query builder.

**Comment by iarmankhan** [comment]
We use objection.js on [EasyCall](https://easycall.io)

**Comment by jstayton** [comment]
We're using it at [Truepic](https://truepic.com)! ❤️

**Comment by caschbre** [comment]
We use it at [Lessonface](https://www.lessonface.com).

**Comment by cesumilo** [comment]
We use it at [Inarix](https://www.inarix.com)! 🚀

**Comment by gamedevsam** [comment]
We use it at [Salesforce](https://www.salesforce.com/), thank you for this amazing ORM library! :heart:

![image](https://user-images.githubusercontent.com/420996/155606637-b5d53f55-a24b-4f5f-8964-62f532c2cb35.png)

**Comment by kapouer** [comment]
I'm using it within https://github.com/pageboard/server and it's been a successful experience.
Looking forward to switch to https://koskimas.github.io/kysely/ eventually !

**Comment by damusix** [comment]
Our company used it for https://zencare.co and https://hybridstrengthcoach.com

**Comment by ralcorta** [comment]
We use it for different projects in Valiulab.com :
https://fundaciongada.org.ar
https://sinconta.com

**Comment by MHamzaRajput** [comment]
www.bkk.ag

**Comment by iSeanWeiDev** [comment]
I am using the Objection.js/Knex.js/DBML with postgres to handle the backend APIs. especially, really good match with Graphql as well. I am using this library since 2018

**Comment by lehni** [comment]
To everybody in this thread, there are some important developments concerning the library, see #2335

**Comment by lehni** [comment]
And I failed to mention this so far, but we're using it at [Lineto.com](https://lineto.com) and are very happy with it.

**Comment by KrisLau** [comment]
> To everybody in this thread, there are some important developments concerning the library, see:
> - #2335

Just to give a quick TLDR to anyone watching this thread, koskimas (current maintainer) is leaving the project and looking for maintainers to take over. If you have enjoyed the library and would like to:

- become an active maintainer 
- have ideas about keeping the library maintained
-  or recruit maintainers
 
Please leave a comment on that thread

**Reply by cesumilo** [reply]
Hi, we are clearly heavy users of objection at Inarix. I can talk with my employer to dedicate some time every month maintaining the project. What would be the process to become active maintainers ? We are two developers interested to become active maintainers.

**Reply by lehni** [reply]
Great news @cesumilo! It looks like I am currently the only active maintainer of the project, very happy to have other people joining.

**Reply by lehni** [reply]
@cesumilo are you still interested in becoming a maintainer? I would love to make this happen, and I do have the permissions to add new collaborators to the repo. There is no standardized process but I think we should start with some current issues that need attending. You could write fixes and submit them as PRs in the normal way, I'd review them, and if the work goees well, we could then discuss the next steps. Sounds good?

**Reply by KrisLau** [reply]
@cesumilo hi just thought I'd check in to see if you're still interested in becoming an active maintainer? the reply above is from the current active maintainer

**Reply by cesumilo** [reply]
Sorry @lehni and @KrisLau, for some reason I didn't have a notification on Github. Yes sure! I've someone else in my team that could help us as well.

@lehni It sounds good to me! Do you have some "good first issue" for me?

I got confirmation from my manager to invest some time every 2 weeks to maintain objection.js on our work time!
Can you add me (@cesumilo) and @yzalium as contributors? Do you have any preferred way to communicate (e.g. Slack, Discord, etc)?

**Reply by yzalium** [reply]
Hi there, will be happy to help!

**Reply by lehni** [reply]
Great news @cesumilo and @yzalium! For issues I would just go through the open issues to see what bug or feature request you feel like is worth tackling (also in light of your company's own needs), and then you can start working on it and submit a PR for review when you're ready. After a few more complex issues solved in this way, we can start discussing becoming full maintainers. Does this sound good? As for good first issues, since I am personally not very involved with TypeScript, I would say that any of those issues are very good one to tackle. At the same time though, I won't be great for reviewing the work :)

**Reply by lehni** [reply]
As for where to discuss, I do use Slack for work but I try to limit its use as much as I can. Perhaps we can try keeping communication to the issues and discussions here before adding a real-time thing to the mix?

**Reply by cesumilo** [reply]
It sounds good to me @lehni 👍 I've raised 2 new PRs and we'll tackle more complex issues slowly by learning new things in the ORM. We'll also continue maintaining the tests.

The idea behind becoming full maintainers is to be able to assign ourselves to issues, to review ourselves on PRs and be more proactive in the planning of future releases using milestones, etc. I don't know if there are other ways to have those abilities?

**Reply by lehni** [reply]
Yes exactly, that's the mid-term goal with becoming full maintainers. But this takes a bit of time I think. I'm also not super confident with all my changes yet, e.g. #2603 and #2605 were caused by my work for #2452 and it took me quite a while to realise that.

**Reply by cesumilo** [reply]
Great! 👌 We'll continue contributing as we're doing right now and you'll tell us when you feel we're ready to become full maintainers.

**Reply by KrisLau** [reply]
I'll continue helping to look for more maintainers!

**Comment by PJvanDiepen** [comment]
We use objection.js at [0-0-0.nl](https://0-0-0.nl/)

**Comment by throrin19** [comment]
We use objection at [Aquassay](https://www.aquassay.com)

**Comment by ram-you** [comment]
At [**MEDIACEPT**](https://mediacept.com/) , we use Objection.js in production on our own website as well as across numerous websites and applications we develop.

**Comment by iSeanWeiDev** [comment]
We use at Arbitexa.com
